### PR TITLE
fix(cli): provide a more robust error message if analysis fails

### PR DESCRIPTION
**Pull request by Zidious**
The use of `--include` and providing an non-existent or invalid CSS selector would not notify user that was the case: 

image
image

Providing an non-existent or invalid CSS selector will now notify the user of said error: 

image

Closes issue: https://github.com/dequelabs/axe-core-npm/issues/207

### CODE REVIEW

File: packages/cli/src/lib/axe-test-urls.ts [hunk 1]
```diff
@@ -91,7 +92,8 @@ const testPages = async (
 
           /* istanbul ignore if */
           if (err) {
-            return reject(err);
+            console.error(error(err));
+            process.exit(1);
```
Comment (stephenmathieson) [inline packages/cli/src/lib/axe-test-urls.ts:96]: Why doesn't `reject(err)` allow the caller to handle the error on their own? I don't understand how/why this fixes anything.

Comment (Zidious) [reply packages/cli/src/lib/axe-test-urls.ts:96]: `reject(err)` doesn't provide a useful error message to the user: 

image

Including `--show-errors` flag does not either: 

image

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: That seems like the problem is how we're handing rejections then.

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: It looks like the issue is here:

https://github.com/dequelabs/axe-core-npm/blob/v4.3.2/packages/cli/src/bin/index.ts#L153

What is `%j \n $s` supposed to do?

Comment (Zidious) [reply packages/cli/src/lib/axe-test-urls.ts:96]: I think the error is here https://github.com/dequelabs/axe-core-npm/blob/v4.3.2/packages/cli/src/lib/axe-test-urls.ts#L87 we use `err: string` so we don't have access to `e.message` or `e.stack` to do the the string formatting you mentioned

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: I see. Yes, we should _only_ be rejecting with `Error`s, not other things. Rejecting/`throw`ing anything but an instance of an `Error` is considered a bad practice and should be avoided. [ESLint protects us from this sort of thing with `throw`](https://eslint.org/docs/rules/no-throw-literal), but doesn't do the same for Promise rejections. We should pass the entire error back, not just a string.

I didn't know about `%j`, thanks for that. FWIW the `Error` class does not have a `.toJSON()`, so I'm pretty sure that's not doing anything.

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: If the types on the line you linked are correct (`err: string`), then we should do something like:

```ts
        axe.analyze((err: Error | string | null, results: AxeResults) => {
          if (config.timer) {
            events?.endTimer('axe-core execution time');
          }

          /* istanbul ignore if */
          if (err) {
            const error = typeof err === 'string' ? new Error(err) : err
            return reject(error);
          }
```

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: It looks like that type is wrong, as we [pass back the whole `Error` (unless in legacy/callback mode)](https://github.com/dequelabs/axe-core-npm/blob/v4.3.2/packages/webdriverjs/src/index.ts#L134).

Comment (Zidious) [reply packages/cli/src/lib/axe-test-urls.ts:96]: Ya, looking at it a bit more, the [type for the callback function](https://github.com/dequelabs/axe-core-npm/blob/v4.3.2/packages/webdriverjs/src/types.ts#L20) should be `Error | null` and not `string | null`. So we can access `e.message` etc. However, your work around does allow the user to _handle_ the error with `--show-errors`

image

Comment (stephenmathieson) [reply packages/cli/src/lib/axe-test-urls.ts:96]: Yes, the type for that function is wrong. IMO it should be changed/corrected.

However, if we want a _very simple fix_ for the issue at hand, we could just fix the `console.error()` call in `src/bin/index.ts`:

```js
    if (!showErrors) {
      console.error(error('An error occurred while testing this page.'));
    } else {
      console.error(error('Error: %s'), e);
    }
```

File: packages/cli/src/lib/axe-test-urls.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import * as WebDriver from 'selenium-webdriver';
 import AxeBuilder from '@axe-core/webdriverjs';
 import { AxeResults } from 'axe-core';
+import { error } from '../lib/utils';
```
Comment (stephenmathieson) [inline packages/cli/src/lib/axe-test-urls.ts:4]:
```suggestion
```

**Review by stephenmathieson — APPROVED** [review]

File: packages/cli/src/lib/axe-test-urls.ts (hunk 2 above)
Comment (Zidious) [reply packages/cli/src/lib/axe-test-urls.ts]: resolved!

File: packages/cli/src/lib/axe-test-urls.ts (hunk 1 above)
Comment (Zidious) [reply packages/cli/src/lib/axe-test-urls.ts]: For reference issue has been raised: https://github.com/dequelabs/axe-core-npm/issues/422

**Review by stephenmathieson — APPROVED** [review]

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
While I support the change, I don't see how the title describes what you changed, or how this closes the issue.

**Pull request merged**